Realm/RLMSyncSubscription.mm

```diff
@@ -0,0 +1,523 @@
+
+- (instancetype)initWithSubscriptionSet:(RLMSyncSubscriptionSet *)subscriptionSet {
+    if (self = [super init]) {
+        _subscriptionSet = subscriptionSet;
+        [_subscriptionSet registerEnumerator:self];
+        return self;
+    }
+    return nil;
+}
+- (NSUInteger)countByEnumeratingWithState:(NSFastEnumerationState *)state
+                                    count:(NSUInteger)len {
+    NSUInteger batchCount = 0, count = [_subscriptionSet count];
+    for (NSUInteger index = state->state; index < count && batchCount < len; ++index) {
+        auto subscription = [_subscriptionSet objectAtIndex:index];
+        _strongBuffer[batchCount] = subscription;
+        batchCount++;
+    }
+
+    for (NSUInteger i = batchCount; i < len; ++i) {
+        _strongBuffer[i] = nil;
+    }
+
+    if (batchCount == 0) {
+        // Release our data if we're done, as we're autoreleased and so may
+        // stick around for a while
+        if (_subscriptionSet) {
+            [_subscriptionSet unregisterEnumerator:self];
+            _subscriptionSet = nil;
+        }
+    }
+
+
+    state->itemsPtr = (__unsafe_unretained id *)(void *)_strongBuffer;
+    state->state += batchCount;
+    state->mutationsPtr = state->extra+1;
+
+    return batchCount;
+}
+
+@end
+
+NSUInteger RLMFastEnumerate(NSFastEnumerationState *state,
+                            NSUInteger len,
+                            RLMSyncSubscriptionSet *collection) {
+    __autoreleasing RLMSyncSubscriptionEnumerator *enumerator;
+    if (state->state == 0) {
+        enumerator = collection.fastEnumerator;
+        state->extra[0] = (long)enumerator;
+        state->extra[1] = collection.count;
+    }
+    else {
+        enumerator = (__bridge id)(void *)state->extra[0];
+    }
+
+    return [enumerator countByEnumeratingWithState:state count:len];
+}
+
+@implementation RLMSyncSubscriptionSet {
+    BOOL _isInWriteTransaction;
+    std::mutex _collectionEnumeratorMutex;
+    RLMRealm *_realm;
+}
+
+- (instancetype)initWithSubscriptionSet:(realm::sync::SubscriptionSet)subscriptionSet
+                                  realm:(RLMRealm *)realm {
+    if (self = [super init]) {
+        _subscriptionSet = std::make_unique<realm::sync::SubscriptionSet>(subscriptionSet);
+        _realm = realm;
+        return self;
+    }
+    return nil;
+}
+
+- (RLMSyncSubscriptionEnumerator *)fastEnumerator {
+    return [[RLMSyncSubscriptionEnumerator alloc] initWithSubscriptionSet:self];
+}
+
+- (void)unregisterEnumerator:(RLMSyncSubscriptionEnumerator *)enumerator {
+    std::lock_guard lock(_collectionEnumeratorMutex);
+    [_enumerators removeObject:enumerator];
+}
+
+- (void)registerEnumerator:(RLMSyncSubscriptionEnumerator *)enumerator {
+    std::lock_guard lock(_collectionEnumeratorMutex);
+    if (!_enumerators) {
+        _enumerators = [NSHashTable hashTableWithOptions:NSPointerFunctionsWeakMemory];
+    }
+    [_enumerators addObject:enumerator];
+}
```

Comment:
There doesn't appear to be any reason for this? RLMRealm keeps track of its active enumerators because it needs to notify them when a write transaction is started, but `_enumerators` is never actually used here.

Reply:
My mistake– I ported over the RLMRealm code assuming that holding onto the enumerators had something to do with holding onto the reference.